src/migmose/mig/reducednestednachrichtenstruktur.py

```diff
@@ -2,16 +2,138 @@
+
+        # Helper function to create a unique identifier for each segment
+        def get_identifier(segment: Optional[NachrichtenstrukturZeile]) -> tuple[str, str]:
+            if segment is None:
+                return ("0", "root")
+            return (segment.zaehler, segment.bezeichnung)
+
+        def count_segments(segment_group: ReducedNestedNachrichtenstruktur | NestedNachrichtenstruktur) -> int:
+            # Start with counting segments directly under the current segment group
+            total_segments = len(segment_group.segmente)
+            # Recursively count segments in nested segment groups
+            for nested_sg in segment_group.segmentgruppen:
+                if nested_sg is not None:
+                    total_segments += count_segments(nested_sg)
+
+            return total_segments
+
+        # Function to process segments and remove duplicates within the same list.
+        def process_segments(segments: list[Optional[NachrichtenstrukturZeile]]):
+            seen = set()
+            unique_segments = []
+            for segment in segments:
+                if segment is not None:
+                    identifier = get_identifier(segment)
+                    if identifier not in seen:
+                        seen.add(identifier)
+                        unique_segments.append(segment)
+            return unique_segments
+
+        # Recursive function to traverse and clean segment groups
+        def process_segmentgruppen(
+            segmentgruppen: list[Optional[NestedNachrichtenstruktur]],
+            segment_count_dict: dict,
+            seen=None,
+            depth: int = 0,
+        ):
+            """Recursively clean segment groups to avoid duplicates, keep largest,
+            with debugging for circular references."""
+            if seen is None:
+                seen = {}
+            result = []
+
+            for sg in segmentgruppen:
+                if sg is not None:
+                    identifier = get_identifier(sg.header_linie)
+                    max_count, max_sg = segment_count_dict[identifier]
+
+                    if identifier not in seen:
+                        seen[identifier] = max_sg
+                        logger.info("Added {} with {} segments at depth {}.", identifier, max_count, depth)
+
+                    sg.segmente = process_segments(max_sg.segmente)
+                    sg.segmentgruppen = process_segmentgruppen(
+                        max_sg.segmentgruppen, segment_count_dict, seen, depth + 1
+                    )
+
+            # Compile the unique list from the seen dictionary after recursive processing to avoid circular reference
+            if depth == 0:  # Only compile on the initial call, not recursive ones
+                result = [seen[key] for key in seen]
+            return result
+
+        def build_segment_count_dict(segment_groups: list[Optional[NestedNachrichtenstruktur]]):
+            segment_count_dict: dict = {}
```

Comment:
```suggestion
            segment_count_dict: dict[tuple[str, str], tuple[int, NestedNachrichtenstruktur]] = {}
```
Bin mir hier gerade aber auch nicht ganz sicher.